{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import genome_helpers as ge\n",
    "import load_genomes \n",
    "import pickle\n",
    "import pandas as pd\n",
    "\n",
    "def pickle_save(filename, variable):\n",
    "    filename = filename+'.pkl'\n",
    "    output = open(filename, 'wb')\n",
    "    pickle.dump(variable, output)\n",
    "    output.close()\n",
    "    print(filename + \" Saved!\")\n",
    "    return True\n",
    "\n",
    "def pickle_open(filename, df=False):\n",
    "    \"\"\"Eventually check if we are opening a python 3 or python 2 file \n",
    "    and deal with discrepencies (python 3 pickle opened in python 2 etc...)\"\"\"\n",
    "    if df == True:\n",
    "        file = pd.read_pickle(filename)\n",
    "        print(filename + \" dataframe sucessfully loaded!\")\n",
    "    else:\n",
    "        file = pickle.load( open( filename, \"rb\" ) )\n",
    "        print(filename + \" sucessfully loaded!\")\n",
    "    return file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## frequencies\n",
    "def codons():\n",
    "    d = {}\n",
    "    for first in ['A','T','G','C']:\n",
    "        for second in ['A','T','G','C']:\n",
    "            for third in ['A','T','G','C']:\n",
    "                d[first+second+third] = 0\n",
    "    return d\n",
    "## depends on nucleotide_frequency, codon_frequency(tri_split aa_frequency(codonToAA codons\n",
    "def nucleotide_frequency_features(table):\n",
    "    res = [[l[1] for l in list(ge.nucleotide_frequency(seg).items())] for seg in list(table.seq)]\n",
    "    frequency_df = pd.DataFrame(res)\n",
    "    frequency_df.columns=[\"a\",\"c\",\"g\",\"t\"]\n",
    "    print(\"Calculated Nucleotide Frequencies...\")\n",
    "    print(frequency_df.head())\n",
    "    return frequency_df\n",
    "def codon_frequency_features(table):\n",
    "    codon_matrix = [list(ge.codon_frequency(ge.tri_split(seg)).values()) for seg in list(table.seq)]\n",
    "    codon_frequency_df = pd.DataFrame(codon_matrix)\n",
    "    codon_frequency_df.columns=[list(codons().keys())]\n",
    "    print(\"Calculated codon frequencies...\")\n",
    "    print(codon_frequency_df.head(2))\n",
    "    return codon_frequency_df\n",
    "def aa_frequency_features(table):\n",
    "    aa_matrix = [list(ge.aa_frequency(ge.codonToAA(ge.tri_split(seg))).values()) for seg in list(table.seq)]\n",
    "    aa_frequency_df = pd.DataFrame(aa_matrix)\n",
    "    aa_frequency_df.columns=[['X','N','T','D','Y','L','W','Q','S','I','M','C','A','V','K','G','H','R','P','F','E']]\n",
    "    print(\"Calculated amino acid frequencies...\")\n",
    "    print(aa_frequency_df.head())\n",
    "    return aa_frequency_df\n",
    "# Forwards\n",
    "# Reverse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "merge_predict_newSeq.pkl dataframe sucessfully loaded!\n"
     ]
    }
   ],
   "source": [
    "merge_predicted = pickle_open('merge_predict_newSeq.pkl',df=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "forward_pairs = merge_predict[merge_predict.direction == 'F']\n",
    "genome_0 = forward_pairs[forward_pairs.genome == 0]\n",
    "all_ys = []\n",
    "all_xs = []\n",
    "for i,pair in enumerate(f_ps):\n",
    "    n_df = nucleotide_frequency_features(pair)\n",
    "    c_df = codon_frequency_features(pair) \n",
    "    aa_df = aa_frequency_features(pair)  \n",
    "    concat_frequencies = pd.concat([n_df,aa_df,c_df],axis=1)\n",
    "    all_xs.append(concat_frequencies)\n",
    "    real_gene = pair.gene\n",
    "    all_ys.append(real_gene)\n",
    "\n",
    "reverse_pairs = merge_predict[merge_predict.direction == 'R']\n",
    "genome_0 = reverse_pairs[reverse_pairs.genome == 0]\n",
    "all_fys = []\n",
    "all_fxs = []\n",
    "\n",
    "def run_reverse_pairs(pair_df):\n",
    "    n_df = nucleotide_frequency_features(pair_df)\n",
    "    c_df = codon_frequency_features(pair_df) \n",
    "    aa_df = aa_frequency_features(pair_df)  \n",
    "    concat_frequencies = pd.concat([n_df,aa_df,c_df],axis=1)\n",
    "    all_fxs.append(concat_frequencies)\n",
    "    real_gene = pair_df.gene\n",
    "    all_fys.append(real_gene)\n",
    "    return [all_fxs,all_fys]\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>direction</th>\n",
       "      <th>frame</th>\n",
       "      <th>genome</th>\n",
       "      <th>new_seq</th>\n",
       "      <th>FeatureScoresList</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>ATGAGACTTATTAAGAATACCACAGAATTAATCGGAATTAAAGACC...</td>\n",
       "      <td>[4.50087732588, 0.926389374604, 3.17156170326,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGTATCAAGAATTAAATGATTATTTAGAATTCCAATCAATGAAGT...</td>\n",
       "      <td>[1.96400923621, 0.313296494764, 2.17883683118,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGGTAACCAAAGTTGAAGAGATTCGCTCATATTTGATAGCTTCTA...</td>\n",
       "      <td>[2.73030519013, 0.890229719259, 3.05451898882,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666</th>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>ATGAAAAATAATAGCAAATCAAAAAGTAATAAGATAAGTCTTTCTC...</td>\n",
       "      <td>[6.74458466691, 0.234255582362, 3.26307411115,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>741</th>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGTCTAAAAAAATTAAGAATTTTATACTATTAAATATTGGATCTC...</td>\n",
       "      <td>[3.12830660119, 0.458363693642, 2.47104929257,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>ATGAATTGGTCTACTATTTGGGAACTTATCAAGATAAACATTCTTT...</td>\n",
       "      <td>[4.5720537119, 0.954723502496, 3.43057583216, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>513</th>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>ATGCTCTGCCAAAACTGTAACCTTAATGAAGCTAGTATCCATTTGT...</td>\n",
       "      <td>[5.93555476826, 2.89222179571, 6.45229127591, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>385</th>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>ATGACAAAAAAAGATGTAATTCAATTATTAGAAAAAATAGCTATAT...</td>\n",
       "      <td>[6.72457773187, 5.0578117881, 4.65344680931, 8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745</th>\n",
       "      <td>True</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>GTGGTTGCTAAAATTTTAGATGGTAAACAAATAGCAAAAGAATATA...</td>\n",
       "      <td>[3.09738583662, 0.385640920527, 2.62133923771,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>ATGAAAGCAATTAAAATTGCGATTGATGGGCCAGCTTCAAGTGGAA...</td>\n",
       "      <td>[1.9717650041, 0.421265032133, 1.21339800872, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene direction  frame  genome  \\\n",
       "411  True         R      0       3   \n",
       "203  True         F      1       1   \n",
       "61   True         F      2       1   \n",
       "666  True         F      0       2   \n",
       "741  True         F      0       1   \n",
       "216  True         R      1       0   \n",
       "513  True         R      0       3   \n",
       "385  True         F      2       2   \n",
       "745  True         R      1       2   \n",
       "358  True         F      1       0   \n",
       "\n",
       "                                               new_seq  \\\n",
       "411  ATGAGACTTATTAAGAATACCACAGAATTAATCGGAATTAAAGACC...   \n",
       "203  ATGTATCAAGAATTAAATGATTATTTAGAATTCCAATCAATGAAGT...   \n",
       "61   ATGGTAACCAAAGTTGAAGAGATTCGCTCATATTTGATAGCTTCTA...   \n",
       "666  ATGAAAAATAATAGCAAATCAAAAAGTAATAAGATAAGTCTTTCTC...   \n",
       "741  ATGTCTAAAAAAATTAAGAATTTTATACTATTAAATATTGGATCTC...   \n",
       "216  ATGAATTGGTCTACTATTTGGGAACTTATCAAGATAAACATTCTTT...   \n",
       "513  ATGCTCTGCCAAAACTGTAACCTTAATGAAGCTAGTATCCATTTGT...   \n",
       "385  ATGACAAAAAAAGATGTAATTCAATTATTAGAAAAAATAGCTATAT...   \n",
       "745  GTGGTTGCTAAAATTTTAGATGGTAAACAAATAGCAAAAGAATATA...   \n",
       "358  ATGAAAGCAATTAAAATTGCGATTGATGGGCCAGCTTCAAGTGGAA...   \n",
       "\n",
       "                                     FeatureScoresList  \n",
       "411  [4.50087732588, 0.926389374604, 3.17156170326,...  \n",
       "203  [1.96400923621, 0.313296494764, 2.17883683118,...  \n",
       "61   [2.73030519013, 0.890229719259, 3.05451898882,...  \n",
       "666  [6.74458466691, 0.234255582362, 3.26307411115,...  \n",
       "741  [3.12830660119, 0.458363693642, 2.47104929257,...  \n",
       "216  [4.5720537119, 0.954723502496, 3.43057583216, ...  \n",
       "513  [5.93555476826, 2.89222179571, 6.45229127591, ...  \n",
       "385  [6.72457773187, 5.0578117881, 4.65344680931, 8...  \n",
       "745  [3.09738583662, 0.385640920527, 2.62133923771,...  \n",
       "358  [1.9717650041, 0.421265032133, 1.21339800872, ...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_predicted.sample(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## for each genome run the logistic regression model\n",
    "## are the gene\n",
    "## A do the frequencies counts etc\n",
    "## improve the model?\n",
    "## Latex the model formula for the slides\n",
    "## how to measure performance? precision, accuracy\n",
    "# http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
